source/server-discovery-and-monitoring/server-monitoring.rst

```diff
@@ -674,13 +675,21 @@ The event API here is assumed to be like the standard `Python Event
             wait()
 
     def setUpConnection():
+        # Take the mutex to avoid a data race becauase this code writes to the connection field and a concurrent
+        # cancelCheck call could be reading from it.
+        mutex.lock()
         connection = new Connection(serverAddress)
         set connection timeout to connectTimeoutMS
-        perform connection handshake
+        mutex.unlock()
+
+        # Do any potentially blocking operations after releasing the mutex.
+        create the socket and perform connection handshake
 
     def checkServer(previousDescription):
         try:
-            if not connection:
+            # The connection is null if this is the first check. It's closed if there was an error during the previous
+            # check or the previous check was cancelled.
+            if not connection or connection.isClosed():
```

Comment:
I think this change is needed because `cancelCheck` closes the connection but does not set it to `null`, so the connection won't be re-created on the next check unless `connection.isClosed` is checked.

Opted to make this change instead of having `cancelCheck` assign the connection to `null` because we'd need to take the lock here too if it did assignment.

Reply:
Acknowledged.

Reply:
LGTM, I already had code like this in the node implementation so I'm glad we're codifying it here